### ISSUE TITLE
Low consumption rate

### ISSUE THREAD

**Issue by pakokol** [eclipse/paho.mqtt.c#1254]
**Description**
I don't know if this is a bug or if I'm doing something wrong. The thing is I tried to use this library to consume messages from the MQTT broker at the rate of about 1000 msg/s. The payload is on the average size of 12KB. To test this I have used the async consumer example and the rate of consumption that I achieved is about 100 msg/s. To note I have removed the part where the payload is printed in the example.

I assumed that this volume of data (1000 msg/s) should be easily digested. To verify this I have tested it with the python implementation of paho as also with the java implementation and in both cases, the consumer could easily consume all the traffic. 

Broker, producer, and consumer are all on the same machine.

**Environment:**
 - OS: Ubuntu 22.04 LTS
 - HW: AMD Ryzen 5, 16GB ram
 - Broker: RabbitMQ
 - Paho.MQTT.C: v1.3.10

**Additional context**
I have also tried the consumer to place on a Windows machine, but the consumption rate was the same.

**Producer code**
```
import threading
import numpy as np

from sparkplug_b import *
from sparkplug_b_pb2 import *
from host_session_establishment import *

number_of_spawned_clients = 1
threads = []
serverUrl = "localhost"
myGroupId = "testGroup"
myNodeName = "testNode"
publishPeriod = 5000
running = True

class AliasMap:
    Next_Server = 0
    Rebirth = 1
    Reboot = 2
    Dataset = 4
    Sine = 5

######################################################################
# The callback for when the client receives a CONNACK response from the server.
######################################################################
def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("Connected with result code "+str(rc))
    else:
        print("Failed to connect with result code "+str(rc))
        sys.exit()

    global myGroupId
    global myNodeName

    # Subscribing in on_connect() means that if we lose the connection and
    # reconnect then subscriptions will be renewed.
    client.subscribe("spBv1.0/" + myGroupId + "/NCMD/" + myNodeName + "/#")
    client.subscribe("spBv1.0/" + myGroupId + "/DCMD/" + myNodeName + "/#")

    # Publish birth for first metric

    columns = ["ACC_X", "ACC_Y", "ACC_Z", "Timestamp"]
    types = [DataSetDataType.Double, DataSetDataType.Double, DataSetDataType.Double, DataSetDataType.UInt64]
    # Periodically publish some new data
    payload = getDdataPayload()
    dataset = initDatasetMetric(payload, "Sine Sample", AliasMap.Sine, columns, types)
    byteArray = bytearray(payload.SerializeToString())
    #client.publish("spBv1.0/" + myGroupId + "/DBIRTH/" + myNodeName + "/test", byteArray, 1, True)


######################################################################

def start_client():
    ######################################################################
    # Main Application
    ######################################################################
    print("Starting main application")
    global running

    # Create the node death payload
    deathPayload = getNodeDeathPayload()

    # Start of main program - Set up the MQTT client connection
    device_id = str(threading.current_thread().name)
    client = mqtt.Client(client_id=device_id, clean_session=False)
    client.on_connect = on_connect

    deathByteArray = bytearray(deathPayload.SerializeToString())
    client.will_set("spBv1.0/" + myGroupId + "/NDEATH/" + myNodeName, deathByteArray, 0, False)
    client.connect(serverUrl, 1883, 20)

    def get_current_utc_in_micro():
        import datetime, time
        nanoseconds = time.time_ns() / 1000
        micro = int(nanoseconds)
        return micro


    # Short delay to allow connect callback to occur
    time.sleep(.1)
    client.loop()

    counter = 0

    time_to_sleep = 1

    length = np.pi * 2
    # Resulution for sending
    resolution = 200
    chunk_number = int(resolution/50)
    my_wave = np.sin(np.arange(0, length, length / resolution))
    columns = ["Timestamp"]
    types = [DataSetDataType.UInt64]
    for i in range(resolution):
        columns.append("v_"+str(i))
        types.append(DataSetDataType.Double)

    def chunks(lst, n):
        """Yield successive n-sized chunks from lst."""
        for i in range(0, len(lst), n):
            l = lst[i:i + n]
            yield np.insert(l, 0, 0)

    chunk_my_wave_list = list(chunks(my_wave, chunk_number))
    print(chunk_my_wave_list)
    chunk_column = ["Timestamp"]
    chunk_types = [DataSetDataType.UInt64]
    for i in range(chunk_number):
        chunk_column.append("s_"+str(i))
        chunk_types.append(DataSetDataType.Double)

    while running:
        for d in chunk_my_wave_list:
            # Periodically publish some new data
            payload = getDdataPayload()
            # Create a DataSet
            dataset = initDatasetMetric(payload, "Sine sample", AliasMap.Sine, chunk_column, chunk_types)
            row = dataset.rows.add()
            element = row.elements.add()
            element.long_value = int(get_current_utc_in_micro())
            for i in range(1, len(d)):
                element = row.elements.add()
                element.double_value = d[i]

            # Publish a message data
            byteArray = bytearray(payload.SerializeToString())
            client.publish("spBv1.0/" + myGroupId + "/DDATA/" + myNodeName + "/" + device_id, byteArray, 1, False)
            # Sit and wait for inbound or outbound events
            #time.sleep(0.01)
            client.loop()
    ######################################################################


for i in range(number_of_spawned_clients):
    threads.append(threading.Thread(target=start_client, name="simulated_nemo_"+str(i)))

# starting threads
for t in threads:
    t.start()

value = input("Enter eny key to stop\n")
print(f'You entered {value}')

running = False

# wait until all threads finish
for t in threads:
    t.join()

```
**Consumer code C**
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "MQTTAsync.h"

#if !defined(_WIN32)
#include <unistd.h>
#else
#include <windows.h>
#endif

#if defined(_WRS_KERNEL)
#include <OsWrapper.h>
#endif

#define ADDRESS     "tcp://localhost:1883"
#define CLIENTID    "ExampleClientSub"
#define TOPIC       "#"
#define PAYLOAD     "Hello World!"
#define QOS         1
#define TIMEOUT     10000L

int disc_finished = 0;
int subscribed = 0;
int finished = 0;

void onConnect(void* context, MQTTAsync_successData* response);
void onConnectFailure(void* context, MQTTAsync_failureData* response);

void connlost(void *context, char *cause)
{
	MQTTAsync client = (MQTTAsync)context;
	MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer;
	int rc;

	printf("\nConnection lost\n");
	if (cause)
		printf("     cause: %s\n", cause);

	printf("Reconnecting\n");
	conn_opts.keepAliveInterval = 20;
	conn_opts.cleansession = 1;
	conn_opts.onSuccess = onConnect;
	conn_opts.onFailure = onConnectFailure;
	if ((rc = MQTTAsync_connect(client, &conn_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start connect, return code %d\n", rc);
		finished = 1;
	}
}


int msgarrvd(void *context, char *topicName, int topicLen, MQTTAsync_message *message)
{
    printf("Message arrived\n");
    printf("     topic: %s\n", topicName);
    // printf("   message: %.*s\n", message->payloadlen, (char*)message->payload);
    MQTTAsync_freeMessage(&message);
    MQTTAsync_free(topicName);
    return 1;
}

void onDisconnectFailure(void* context, MQTTAsync_failureData* response)
{
	printf("Disconnect failed, rc %d\n", response->code);
	disc_finished = 1;
}

void onDisconnect(void* context, MQTTAsync_successData* response)
{
	printf("Successful disconnection\n");
	disc_finished = 1;
}

void onSubscribe(void* context, MQTTAsync_successData* response)
{
	printf("Subscribe succeeded\n");
	subscribed = 1;
}

void onSubscribeFailure(void* context, MQTTAsync_failureData* response)
{
	printf("Subscribe failed, rc %d\n", response->code);
	finished = 1;
}


void onConnectFailure(void* context, MQTTAsync_failureData* response)
{
	printf("Connect failed, rc %d\n", response->code);
	finished = 1;
}


void onConnect(void* context, MQTTAsync_successData* response)
{
	MQTTAsync client = (MQTTAsync)context;
	MQTTAsync_responseOptions opts = MQTTAsync_responseOptions_initializer;
	int rc;

	printf("Successful connection\n");

	printf("Subscribing to topic %s\nfor client %s using QoS%d\n\n"
           "Press Q<Enter> to quit\n\n", TOPIC, CLIENTID, QOS);
	opts.onSuccess = onSubscribe;
	opts.onFailure = onSubscribeFailure;
	opts.context = client;
	if ((rc = MQTTAsync_subscribe(client, TOPIC, QOS, &opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start subscribe, return code %d\n", rc);
		finished = 1;
	}
}


int main(int argc, char* argv[])
{
	MQTTAsync client;
	MQTTAsync_connectOptions conn_opts = MQTTAsync_connectOptions_initializer;
	MQTTAsync_disconnectOptions disc_opts = MQTTAsync_disconnectOptions_initializer;
	int rc;
	int ch;

	if ((rc = MQTTAsync_create(&client, ADDRESS, CLIENTID, MQTTCLIENT_PERSISTENCE_NONE, NULL))
			!= MQTTASYNC_SUCCESS)
	{
		printf("Failed to create client, return code %d\n", rc);
		rc = EXIT_FAILURE;
		goto exit;
	}

	if ((rc = MQTTAsync_setCallbacks(client, client, connlost, msgarrvd, NULL)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to set callbacks, return code %d\n", rc);
		rc = EXIT_FAILURE;
		goto destroy_exit;
	}

	conn_opts.keepAliveInterval = 20;
	conn_opts.cleansession = 1;
	conn_opts.onSuccess = onConnect;
	conn_opts.onFailure = onConnectFailure;
	conn_opts.context = client;
	if ((rc = MQTTAsync_connect(client, &conn_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start connect, return code %d\n", rc);
		rc = EXIT_FAILURE;
		goto destroy_exit;
	}

	while (!subscribed && !finished)
		#if defined(_WIN32)
			Sleep(100);
		#else
			usleep(10000L);
		#endif

	if (finished)
		goto exit;

	do 
	{
		ch = getchar();
	} while (ch!='Q' && ch != 'q');

	disc_opts.onSuccess = onDisconnect;
	disc_opts.onFailure = onDisconnectFailure;
	if ((rc = MQTTAsync_disconnect(client, &disc_opts)) != MQTTASYNC_SUCCESS)
	{
		printf("Failed to start disconnect, return code %d\n", rc);
		rc = EXIT_FAILURE;
		goto destroy_exit;
	}
 	while (!disc_finished)
 	{
		#if defined(_WIN32)
			Sleep(100);
		#else
			usleep(10000L);
		#endif
 	}

destroy_exit:
	MQTTAsync_destroy(&client);
exit:
 	return rc;
}
```

**Consumer code Python**
```
from sparkplug_b_pb2 import *
from host_session_establishment import *

serverUrl = "localhost"
running = True


def on_connect(client, userdata, flags, rc):
    if rc == 0:
        print("Connected with result code "+str(rc))
    else:
        print("Failed to connect with result code "+str(rc))
        sys.exit()

    # Subscribing in on_connect() means that if we lose the connection and
    # reconnect then subscriptions will be renewed.
    client.subscribe("spBv1.0/#", qos=1)
    client.subscribe("test", qos=1)


def on_message(client, userdata, message):
    pass


client = mqtt.Client(client_id="test", clean_session=False)
client.on_connect = on_connect
client.on_message = on_message

client.connect(serverUrl, 1883, 20)
client.loop_start()

while running:
    time.sleep(0.1)
```

**Comment by pakokol** [comment]
I have also tested with few other versions of Paho Mqtt C library and it seems that this issue impacts only the version 1.3.10. The following versions that I have tested are: 1.3.9, 1.3.8, 1.3.7 and 1.3.6.

**Comment by icraggs** [comment]
I put a change in the develop branch which will affect/fix this.

**Comment by pakokol** [comment]
Thank you for your fast reply! I will test it tomorrow and let you know the results.

**Comment by pakokol** [comment]
Hi Ian! I have conducted further tests and it seems that the issue resolves on Linux systems, but on the Windows platform, it still exists. I have also checked the classic version of the library and the issue there doesn't show up.

**Comment by fpagliughi** [comment]
I just hit this when doing performance tests with the Paho Rust wrapper. So glad to hear there's already a fix under way! @icraggs let me know if there's anything I can help with to get this fully tested and released. I will start testing with the Rust client this weekend.

**Comment by icraggs** [comment]
On Windows with the master branch (1.3.10) I get 10 msgs per second, with the develop branch I get 3232 msgs per second. So I think it's fixed for me on Windows, or at least some of it is. This is the application I used.

[measureloop.txt](https://github.com/eclipse/paho.mqtt.c/files/9275239/measureloop.txt)

**Comment by fpagliughi** [comment]
I can confirm the `develop` branch fixes the issue I was seeing on Linux.

I have several Rust services that do RPC's to each other over MQTT v5 (using ResponseTopic and CorrelationData properties, etc). I wrote a performance monitor to watch them and was shocked to see that all requests were consistently taking over 100ms to execute. I was about to investigate when I saw this issue.

With the `develop` branch I'm seeing 0.4 - 2ms. That makes more sense.

I would say this this is important enough for an update/release of the C and Rust libraries. (The C++ can work with any compatible C lib, so would get the fix automatically).

**Comment by icraggs** [comment]
I was waiting for a comment about Windows. I think it's fixed there too.

**Comment by fpagliughi** [comment]
I'm just back from holiday and can start testing on Windows this week, to see what I can see.

**Comment by fpagliughi** [comment]
Apologies for the delay on this... but I tested on Windows, and it appears the problem is fixed there as well, by my tests.

As I mentioned, I have some Rust services using the request/response RPC pattern. 
With the released 1.3.10 branch, I'm seeing round-trip times of ~50-100ms (time from request message sent to reply message received). 
With the develop branch, I'm seeing ~5ms.

Sorry that it's not a direct measurement, but it's a very practical one for my usage.

Much better. Thanks, @icraggs .
